src/server/lock_manager/deadlock.rs

```diff
@@ -390,20 +383,99 @@ impl Scheduler {
     pub fn validate(&self, f: Box<dyn FnOnce(u64) + Send>) {
         self.notify_scheduler(Task::Validate(f));
     }
+
+    #[cfg(test)]
+    pub fn get_role(&self, f: Box<dyn FnOnce(Role) + Send>) {
+        self.notify_scheduler(Task::GetRole(f));
+    }
 }
 
-impl Coprocessor for Scheduler {}
+/// The leader region is the region containing the LEADER_KEY and the leader of the
+/// leader region is also the leader of the deadlock detector.
+const LEADER_KEY: &[u8] = b"";
 
-/// Implements observer traits for `Scheduler`.
-/// If the role of the node in the leader region changes, notifys the deadlock detector.
-///
-/// If the leader region is merged or splited in the node, the role of the node won't change.
-/// If the leader region is removed and the node is the leader, it will change to follower first.
-/// So there is no need to observe region change events.
-impl RoleObserver for Scheduler {
+/// `RoleChangeNotifier` observes region or role change events of raftstore. If the
+/// region is the leader region and the role of this node is changed, a `ChangeRole`
+/// task will be scheduled to the deadlock detector. It's the only way to change the
+/// node from the leader of deadlock detector to follower, and vice versa.
+#[derive(Clone)]
+pub(crate) struct RoleChangeNotifier {
+    /// The id of the valid leader region.
+    // raftstore.coprocessor needs it to be Sync + Send.
+    leader_region_id: Arc<Mutex<u64>>,
+    scheduler: Scheduler,
+}
+
+impl RoleChangeNotifier {
+    fn is_leader_region(region: &Region) -> bool {
+        // The key range of a new created region is empty which misleads the leader
+        // of the deadlock detector stepping down.
+        //
+        // If the peers of a region is not empty, the region info is complete.
+        !region.get_peers().is_empty()
+            && region.get_start_key() <= LEADER_KEY
+            && (region.get_end_key().is_empty() || LEADER_KEY < region.get_end_key())
+    }
+
+    pub(crate) fn new(scheduler: Scheduler) -> Self {
+        Self {
+            leader_region_id: Arc::new(Mutex::new(INVALID_ID)),
+            scheduler,
+        }
+    }
+
+    pub(crate) fn register(self, host: &mut CoprocessorHost) {
+        host.registry
+            .register_role_observer(1, Box::new(self.clone()));
+        host.registry
+            .register_region_change_observer(1, Box::new(self));
+    }
+}
+
+impl Coprocessor for RoleChangeNotifier {}
+
+impl RoleObserver for RoleChangeNotifier {
     fn on_role_change(&self, ctx: &mut ObserverContext<'_>, role: StateRole) {
-        if is_leader_region(ctx.region()) {
-            self.change_role(role.into());
+        let region = ctx.region();
+        // A region is created first, so the leader region id must be valid.
```

Comment:
if `is_leader_region` is false, please return as early as possible.

Reply:
It is.